### DISCUSSION TITLE
Releasing Unused Controls

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#9181] [Q&A]
Working on MvvmDialogs for single-page navigation, my idea is to store a navigation history of ViewModels, and maintain a cache of weak references to Views. The idea is that only a single View needs to be maintained in memory, and when you browse to other pages, the views can be released.

Upon testing, however, I'm unable to get the views to garbage collect. [This test fails.](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/blob/master/src/MvvmDialogs.Avalonia.Tests/NavigationTests.cs#L104)

Code isn't too complex, [NavigationManager](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/blob/master/src/MvvmDialogs.Avalonia/Navigation/NavigationManager.cs) does most of the job, [ViewCache](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/blob/master/src/MvvmDialogs.Avalonia/Navigation/ViewCache.cs) keeps a cache of weak references.

I really can't find what would keep a reference to the first View once the second has been shown.

My question is... is there something within the Avalonia framework that could preserve a reference and block garbage collection?

**Comment by BAndysc** [comment]
Sometimes Avalonia keeps unnecessary references, but in your case, I believe this object can't be garbage collected due to your code. This object has to be _valid_ during the whole lifetime of this function, try to move
```
var vm1 = new FirstViewModel();
DialogService.Show(null, vm1);
var view1 = new WeakReference(NavigationManager.CurrentView);
```
to a separate function, so that `vm1` will not have valid references in your code.

If it still doesn't help, I would suggest using a memory profiler like https://www.jetbrains.com/dotmemory/ (there is a 30 day trial), which will precisely tell you what are the references to the object (= what objects prevent your object from being garbage collected)

**Reply by mysteryx93** [reply]
vm1 is the ViewModel, not the View, so that should be fine.

I have JetBrains Rider... been trying to figure out how to launch dotMemory profiler in the latest version... haven't figured it out yet

**Reply by BAndysc** [reply]
oops, my bad, I though you are checking if the ViewModel still exists. So yeah, I recommend dotMemory, run in under it and then you have to find the view Instance and check its shortest paths to root https://www.youtube.com/watch?v=igIpUE1KhX4
I am not sure how to do it directly from the Rider

**Reply by mysteryx93** [reply]
dotMemory has gone through a lot of changes so most of the guides do not reflect the current version. I'll give it another try later to see if I can figure it out.
